### PR TITLE
Add random and permuted corruption methods

### PR DESCRIPTION
**Pull request by arthurPignet**
Add two corruption methods.

Permuted :  a permutation matrix is randomly generated, stored, and used to flip the labels
Random : a Dirichlet distribution is generated for each label, stored in a stochastic matrix, and used to flip the labels. 

Signed-off-by: arthurPignet <arthur.pignet@mines-paristech.fr>

### CODE REVIEW

File: mplc/partner.py [hunk 1]
```diff
@@ -49,6 +55,38 @@ def corrupt_labels(self, proportion_corrupted):
...
+    def random_labels(self, proportion_corrupted=1):
+        if not 0 <= proportion_corrupted <= 1:
+            raise ValueError(
+                f"The proportion of labels to corrupted was {proportion_corrupted} but it must be between 0 and 1."
+            )
+        # Select the indices where the label will be off-set
+        n = int(len(self.y_train) * proportion_corrupted)
+        idx = sample(list(range(len(self.y_train))), n)
+        # Generate the random matrix to use
+        alpha = np.random.random(self.num_labels)
```
Comment (Thomas-Galtier) [inline mplc/partner.py:82]: It is better to set the parameters to ones in order to have an uniform distribution
```suggestion
        alpha = np.ones(self.num_labels)
```

**Review by Thomas-Galtier — CHANGES_REQUESTED** [review]

File: tests/unit_tests.py [hunk 2]
```diff
@@ -180,33 +190,41 @@ def test_raiseException(self, create_Scenario):
...
+        assert (partner.corruption_matrix.sum(axis=1).round(1) == ones_vect).all()
+
+    def test_shuffle_labels(self, create_Partner):
+        """partner.y_train should be a numpy.ndarray"""
+        partner = create_Partner
+        one_label = np.argmax(partner.y_train[-1])
+        partner.corrupt_labels(1.)
+        assert partner.y_train[-1].max() == 1
+        assert partner.y_train[-1].sum() == 1
+        assert one_label != np.argmax(partner.y_train[-1])
```
Comment (Thomas-Galtier) [inline tests/unit_tests.py:227]: This last test should raise an error one out of ten times. Ones should modify shuffled to ensure it is not the case.

**Review by Thomas-Galtier — CHANGES_REQUESTED** [review]

File: tests/unit_tests.py [hunk 3]
```diff
@@ -180,33 +190,41 @@ def test_raiseException(self, create_Scenario):
...
+        assert partner.y_train[-1].max() == 1
+        assert partner.y_train[-1].sum() == 1
+        ones_vect = np.ones(partner.y_train.shape[1])
+        assert (partner.corruption_matrix.sum(axis=1).round(1) == ones_vect).all()
+
+    def test_shuffle_labels(self, create_Partner):
+        """partner.y_train should be a numpy.ndarray"""
+        partner = create_Partner
+        one_label = np.argmax(partner.y_train[-1])
+        partner.corrupt_labels(1.)
```
Comment (Thomas-Galtier) [inline tests/unit_tests.py:224]:
```suggestion
        partner.shuffle_labels(1.)
```

File: tests/unit_tests.py (hunk 2 above)
Comment (arthurPignet) [reply tests/unit_tests.py:227]: By changing the shuffle method we create a more complex distribution of labels, I think I would remove this test

File: tests/unit_tests.py (hunk 3 above)
Comment (arthurPignet) [reply tests/unit_tests.py:224]: 👍

File: mplc/scenario.py [hunk 4]
```diff
@@ -733,27 +733,53 @@ def data_corruption(self):
...
             elif self.corrupted_datasets[partner_index] == "shuffled":
                 logger.debug(
                     f"   ... Corrupting (by shuffling labels) 100.0 \
-                    percent of the data of partner #{partner.id}"
+                            percent of the data of partner #{partner.id}"
                 )
                 partner.shuffle_labels(1.0)
+            elif self.corrupted_datasets[partner_index] == 'permuted':
+                logger.debug(
+                    f'  ... Corrupting (by permuting the labels ) 100.0 percent of the data of partner #{partner.id}.'
```
Comment (bowni) [inline mplc/scenario.py:747]:
```suggestion
                    f'  ... Corrupting (by permuting the labels) the whole dataset of partner #{partner.id}.'
```

Comment (bowni) [reply mplc/scenario.py:747]: And same for the 4 cases

File: mplc/scenario.py [hunk 5]
```diff
@@ -733,27 +733,53 @@ def data_corruption(self):
...
                     f"   ... Corrupting (by shuffling labels) {self.corrupted_datasets[partner_index][1] * 100} \
-                    percent of the data of partner #{partner.id}"
+                            percent of the data of partner #{partner.id}"
                 )
                 partner.shuffle_labels(self.corrupted_datasets[partner_index][1])
+            elif self.corrupted_datasets[partner_index][0] == 'permuted':
+                logger.debug(
+                    f'  ... Corrupting (by permuting the labels ) {self.corrupted_datasets[partner_index][1] * 100} '
+                    f'percent of the data of partner #{partner.id}.'
+                    f'Permutation matrix available'
```
Comment (bowni) [inline mplc/scenario.py:773]: I think that double-quotes are recommended

File: mplc/partner.py [hunk 6]
```diff
@@ -49,6 +77,42 @@ def corrupt_labels(self, proportion_corrupted):
...
             self.y_train[i] = new_label
 
+    @_Decorator.categorical_needed
+    def permute_labels(self, proportion_corrupted=1):
+        if not 0 <= proportion_corrupted <= 1:
+            raise ValueError(
+                f"The proportion of labels to corrupted was {proportion_corrupted} but it must be between 0 and 1."
+            )
+
+        # Select the indices where the label will be off-set
```
Comment (bowni) [inline mplc/partner.py:87]:
```suggestion
        # Select the indices where the label will be permuted
```

**Review by bowni — APPROVED** [review]
Seems OK to me. Well done to introduce decorators, we are not very used to it yet.

Note: I haven't checked Travis failure

**Review by Thomas-Galtier — APPROVED** [review]

**Pull request merged**